Source/ObjectiveDropboxOfficial/Shared/Handwritten/Networking/DBGlobalErrorResponseHandler.m

```diff
@@ -0,0 +1,187 @@
+///
+/// Copyright (c) 2016 Dropbox, Inc. All rights reserved.
+///
+
+#import "DBGlobalErrorResponseHandler.h"
+#import "DBRequestErrors.h"
+#import "DBTransportBaseClient+Internal.h"
+#import <objc/runtime.h>
+
+static DBNetworkErrorResponseBlock s_networkErrorResponseBlock = nil;
+static NSOperationQueue *s_networkErrorQueue;
+
+static NSMutableDictionary<Class, DBRouteErrorResponseBlock> * _Nullable s_routeErrorToResponseBlock;
+static NSMutableDictionary<Class, NSOperationQueue *> * _Nullable s_routeErrorToQueue;
+
+@implementation DBGlobalErrorResponseHandler
+
++ (void)initialize {
+  static dispatch_once_t once;
+  dispatch_once(&once, ^{
+    s_networkErrorQueue = [NSOperationQueue mainQueue];
+
+    s_routeErrorToResponseBlock = [NSMutableDictionary new];
+    s_routeErrorToQueue = [NSMutableDictionary new];
+  });
+}
+
++ (void)registerRouteErrorResponseBlock:(DBRouteErrorResponseBlock)routeResponseBlock
+                         routeErrorType:(id)routeErrorType {
+  [self registerRouteErrorResponseBlock:routeResponseBlock routeErrorType:routeErrorType queue:nil];
+}
+
++ (void)registerRouteErrorResponseBlock:(DBRouteErrorResponseBlock)routeResponseBlock
+                         routeErrorType:(id)routeErrorType
+                                  queue:(NSOperationQueue *)queue {
+  NSOperationQueue *queueToUse = queue ?: [NSOperationQueue mainQueue];
+
+  @synchronized([DBGlobalErrorResponseHandler class]) {
+    s_routeErrorToResponseBlock[routeErrorType] = routeResponseBlock;
+    s_routeErrorToQueue[routeErrorType] = queueToUse;
+  }
+}
+
++ (void)removeRouteErrorResponseBlockWithRouteErrorType:(id)routeErrorType {
+  @synchronized([DBGlobalErrorResponseHandler class]) {
+    [s_routeErrorToResponseBlock removeObjectForKey:routeErrorType];
+    [s_routeErrorToQueue removeObjectForKey:routeErrorType];
+  }
+}
+
++ (void)registerNetworkErrorResponseBlock:(DBNetworkErrorResponseBlock)networkErrorResponseBlock {
+  [self registerNetworkErrorResponseBlock:networkErrorResponseBlock queue:nil];
+}
+
++ (void)registerNetworkErrorResponseBlock:(DBNetworkErrorResponseBlock)networkErrorResponseBlock
+                                    queue:(NSOperationQueue * _Nullable)queue {
+  NSOperationQueue *queueToUse = queue;
+
+  @synchronized([DBGlobalErrorResponseHandler class]) {
+    s_networkErrorResponseBlock = networkErrorResponseBlock;
+    if (queueToUse) {
+      s_networkErrorQueue = queueToUse;
```

Comment:
This has a similar problem as the last diff, where if you call `registerNetworkErrorResponseBlock:queue` twice and provide a queue the first time but not the second time, the first queue will stick around.